### PR TITLE
Porting ToolStrip PerMonitorV2 HighDpi changes from classic framework

### LINKED ISSUE #215: Port ToolStrip HighDpi improvements from Framework runtime to Core
This is a back-port of the bug fix from .NET 4.8 (BUG 378542) and its original description:

Adjustments for controls for smooth operation in HighDPI PerMonitorV2 mode are usually straight-forward, since the basic infrastructure for this is already implemented in the control class.
Basically, the problem has to be solved that if a form is moved from one monitor to another with a different pixel density, the size of the form and also the size of all controls placed on it have to be adapted to the new resolution.

One starting point here is the fact that the font size of the form is adjusted according to the new Monitor DPI value. Since the font property is an ambient property, the font size changes propagate to all hosted controls so that they are also affected by the font size changes, and for which the display has already been adapted to the changed pixel density. The trigger for the font adjustment is the method `WmDpiChangedBeforeParent` in the control class. 
 
Secondly, further DPI-dependent adjustments must be made, and the `RescaleConstantsForDpi` method, which can be overwritten, provides the second approach.
 
**Challenges with ToolStrip based Controls**
 
All classes inherited from `ToolStripItem` are not controls, but components, so they are not included in the normal method inheritance chain, or in other words: `RescaleConstantsForDpi` is not invoked for component based classes. `ToolStripDropDown`, which forms the container for an opened pulldown menu/pulldown `ToolBoxButton`, is only instantiated if needed and is not part of the normal control tree of a Forms/Control container. For this reason, this control does not implicitly gets its overridden `RescaleConstantsForDPI` called either.
 
**Solution**
 
In order to get the right point in time within the normally excluded control classes when the DPI change (the transition from one monitor to another) occurs, we enable `ToolStrip` itself in a way other controls/components can register at the `ToolStrip` to receive a kind of `Rescale` event. This is done with the new field variable `rescaleConstsCallbackDelegate`. Because `ToolStrip` itself is placed on a form or container, its `RescaleConstantsForDPI` is called accordingly, so it performs the Invoke based on this delegate, triggering all listening components. However, this event must be initially wired for all of the `ToolStripItem`-based components it is hosting, and this is done in `ToolStripItem`'s `SetOwner` method.
This means that `ToolStrip`, all controls derived from `ToolStrip` and also all components derived from `ToolStripItem` are triggered when the DPI changes and the required size adjustments are due. A problem remains with the `ToolStripDropDown` containers, which host only `ToolStripDropDownItem`-based components, but which again are derived from `ToolStripItem`. The latter enables us to intervene in the rescaling event, determine if the affected `ToolStripItem` "has" a `DropDown`, and then make any necessary rescaling here (and in any subordinate) as well. This happens in the `ProcessDropDownItems` method of the `ToolStripDropDownItem` class, which in turn is called by `ToolStrip_RescaleConstants` (and which in turn is triggered by the new delegate in `ToolStrip` and the wiring of `ToolStripItem`).
 
Code changes in all remaining classes then serve as far as possible to make the actual size adjustments for the new DPI value.

### PR DESCRIPTION
**Pull request by KlausLoeffelmann**
fixes #215

This PR is a port of the basic ToolStrip PerMonitorV2 functionality from the classic framework, but without taking the HighDpi Quirks from the classic runtime into account. It addresses those ToolStrip HighDpi issues which were needed at the time to make Visual Studio render ToolStrip based items in PMV2 correctly.

Basically, the problem has to be solved that if a form is moved from one monitor to another with a different pixel density, the size of the form and also the size of all controls placed on it have to be adapted to the new resolution. 

In order to get the right point in time within the normally excluded control classes when the DPI change (the transition from one monitor to another) occurs, we enable ToolStrip itself in a way other controls/components can register at the ToolStrip to receive a kind of Rescale event. This is done with the new field variable rescaleConstsCallbackDelegate. Because ToolStrip itself is placed on a form or container, its RescaleConstantsForDPI is called accordingly, so it performs the Invoke based on this delegate, triggering all listening components. However, this event must be initially wired for all of the ToolStripItem-based components it is hosting, and this is done in ToolStripItem's SetOwner method.

This means that ToolStrip, all controls derived from ToolStrip and also all components derived from ToolStripItem are triggered when the DPI changes and the required size adjustments are due. A problem remains with the ToolStripDropDown containers, which host only ToolStripDropDownItem-based components, but which again are derived from ToolStripItem. The latter enables us to intervene in the rescaling event, determine if the affected ToolStripItem "has" a DropDown , and then make any necessary rescaling here (and in any subordinate) as well. This happens in the ProcessDropDownItems method of the ToolStripDropDownItem class, which in turn is called by ToolStrip_RescaleConstants (and which in turn is triggered by the new delegate in ToolStrip and the wiring of ToolStripItem).

Code changes in all remaining classes then serve as far as possible to make the actual size adjustments for the new DPI value.

### CODE REVIEW

File: src/Common/src/DpiHelper.cs [hunk 1]
```diff
@@ -96,6 +96,15 @@ internal static void InitializeDpiHelperForWinforms()
             isInitializeDpiHelperForWinforms = true;
         }
 
+        // TODO: We need to talk about this.
+        internal static bool DoesCurrentContextRequireScaling
+        {
+            get
```
Comment (JuditRose) [inline src/Common/src/DpiHelper.cs]:
```suggestion
            get => true;
```

File: src/System.Windows.Forms/src/System/Windows/Forms/MenuStrip.cs [hunk 2]
```diff
@@ -73,14 +73,17 @@ public class MenuStrip : ToolStrip {
             get {
                 // MenuStrip control is scaled by Control::ScaleControl()
                 // Ensure grip aligns properly when set visible.
-                return new Padding(2, 2, 0, 2);
-            }
+                return DpiHelper.IsPerMonitorV2Awareness ? 
+                       DpiHelper.LogicalToDeviceUnits(new Padding(2, 2, 0, 2), DeviceDpi) :
+                       new Padding(2, 2, 0, 2);            }
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/MenuStrip.cs]:
```suggestion
                       new Padding(2, 2, 0, 2);
            }
```

File: src/System.Windows.Forms/src/System/Windows/Forms/MenuStrip.cs [hunk 3]
```diff
@@ -73,14 +73,17 @@ public class MenuStrip : ToolStrip {
...
         }
 
         /// <include file='doc\MenuStrip.uex' path='docs/doc[@for="MenuStrip.DefaultSize"]/*' />
         protected override Size DefaultSize {
             get {
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/MenuStrip.cs]:
```suggestion
            get  => DpiHelper.IsPerMonitorV2Awareness ?
                       DpiHelper.LogicalToDeviceUnits(new Size(200, 24), DeviceDpi) :
                       new Size(200, 24);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs [hunk 4]
```diff
@@ -635,7 +644,9 @@ public class ToolStrip : System.Windows.Forms.ScrollableControl,
         /// </devdoc>
         protected override Size DefaultSize {
             get {
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs]:
```suggestion
             get => DpiHelper.IsPerMonitorV2Awareness ?
                       DpiHelper.LogicalToDeviceUnits(new Size(100, 25), DeviceDpi) :
                       new Size(100, 25);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripButton.cs [hunk 5]
```diff
@@ -177,6 +177,22 @@ public class ToolStripButton : ToolStripItem {
             }
         }
 
+        internal override int DeviceDpi {
+            get {
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripButton.cs]:
```suggestion
            get => base.DeviceDpi;
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs [hunk 6]
```diff
@@ -484,6 +486,17 @@ public abstract class ToolStripItem : Component,
             }
         }
 
+        // Every ToolStripItem needs to cache its last/current Parent's DeviceDpi 
+        // for PerMonitorV2 scaling purposes.
+        internal virtual int DeviceDpi {
+            get {
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs]:
```suggestion
            get => deviceDpi;
            set => deviceDpi = value;
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs [hunk 7]
```diff
@@ -484,6 +486,17 @@ public abstract class ToolStripItem : Component,
...
+                return deviceDpi;
+            }
+            set {
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs]:
```suggestion
            set => deviceDpi = value;
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs [hunk 8]
```diff
@@ -710,7 +723,9 @@ public abstract class ToolStripItem : Component,
         /// </devdoc>
         protected virtual Size DefaultSize {
             get {
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripItem.cs]:
```suggestion
            get => return DpiHelper.IsPerMonitorV2Awareness ?
                       DpiHelper.LogicalToDeviceUnits(new Size(23, 23), DeviceDpi) :
                       new Size(23, 23);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripMenuItem.cs [hunk 9]
```diff
@@ -191,11 +191,12 @@ public class ToolStripMenuItem : ToolStripDropDownItem {
         /// </devdoc>
         protected override Size DefaultSize {
             get {
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripMenuItem.cs]:
```suggestion
             get => DpiHelper.IsPerMonitorV2Awareness ?
                      DpiHelper.LogicalToDeviceUnits(new Size(32, 19), DeviceDpi) :
                      new Size(32, 19);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripMenuItem.cs [hunk 10]
```diff
@@ -661,6 +662,24 @@ private static Bitmap GetBitmapFromIcon(string iconName, Size desiredIconSize)
             return menuItem;
        }
 
+        internal override int DeviceDpi
+        {
+            get
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripMenuItem.cs]:
```suggestion
            get => base.DeviceDpi;
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs [hunk 11]
```diff
@@ -174,21 +202,26 @@ public abstract class ToolStripDropDownItem : ToolStripItem {
         SRCategory(nameof(SR.CatAction)),
         SRDescription(nameof(SR.ToolStripDropDownClosedDecr))
         ]
-        public event EventHandler DropDownClosed {
-            add {
+        public event EventHandler DropDownClosed
+        {
+            add
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs]:
```suggestion
            add => Events.AddHandler(EventDropDownClosed, value);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs [hunk 12]
```diff
@@ -174,21 +202,26 @@ public abstract class ToolStripDropDownItem : ToolStripItem {
...
+            {
                 Events.AddHandler(EventDropDownClosed, value);
             }
-            remove {
+            remove
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs]:
```suggestion
            remove => Events.RemoveHandler(EventDropDownClosed, value);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs [hunk 13]
```diff
@@ -201,14 +234,17 @@ internal protected virtual Point DropDownLocation
         SRCategory(nameof(SR.CatAction)),
         SRDescription(nameof(SR.ToolStripDropDownOpeningDescr))
         ]
-        public event EventHandler DropDownOpening {
-            add {
+        public event EventHandler DropDownOpening
+        {
+            add
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs]:
```suggestion
            add => Events.AddHandler(EventDropDownShow, value);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs [hunk 14]
```diff
@@ -201,14 +234,17 @@ internal protected virtual Point DropDownLocation
...
+            {
                 Events.AddHandler(EventDropDownShow, value);
             }
-            remove {
+            remove
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs]:
```suggestion
            remove => Events.RemoveHandler(EventDropDownShow, value);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs [hunk 15]
```diff
@@ -217,14 +253,17 @@ internal protected virtual Point DropDownLocation
         SRCategory(nameof(SR.CatAction)),
         SRDescription(nameof(SR.ToolStripDropDownOpenedDescr))
         ]
-        public event EventHandler DropDownOpened {
-            add {
+        public event EventHandler DropDownOpened
+        {
+            add
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs]:
```suggestion
            add => Events.AddHandler(EventDropDownOpened, value);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs [hunk 16]
```diff
@@ -217,14 +253,17 @@ internal protected virtual Point DropDownLocation
...
+            {
                 Events.AddHandler(EventDropDownOpened, value);
             }
-            remove {
+            remove
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs]:
```suggestion
            remove => Events.RemoveHandler(EventDropDownOpened, value);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs [hunk 17]
```diff
@@ -235,8 +274,10 @@ internal protected virtual Point DropDownLocation
         SRCategory(nameof(SR.CatData)),
         SRDescription(nameof(SR.ToolStripDropDownItemsDescr))
         ]
-        public ToolStripItemCollection DropDownItems {
-            get {
+        public ToolStripItemCollection DropDownItems
+        {
+            get
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs]:
```suggestion
            get => return DropDown.Items;
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs [hunk 18]
```diff
@@ -246,111 +287,131 @@ internal protected virtual Point DropDownLocation
         /// Occurs when the dropdown is opened
         /// </devdoc>
         [SRCategory(nameof(SR.CatAction))]
-        public event ToolStripItemClickedEventHandler DropDownItemClicked {
-            add {
+        public event ToolStripItemClickedEventHandler DropDownItemClicked
+        {
+            add
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs]:
```suggestion
            add => Events.AddHandler(EventDropDownItemClicked, value);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs [hunk 19]
```diff
@@ -246,111 +287,131 @@ internal protected virtual Point DropDownLocation
...
+            {
                 Events.AddHandler(EventDropDownItemClicked, value);
             }
-            remove {
+            remove
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs]:
```suggestion
            remove => Events.RemoveHandler(EventDropDownItemClicked, value);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs [hunk 20]
```diff
@@ -246,111 +287,131 @@ internal protected virtual Point DropDownLocation
...
             }
         }
 
         /// <include file='doc\ToolStripPopupItem.uex' path='docs/doc[@for="ToolStripDropDownItem.HasDropDownItems"]/*' />
         [Browsable(false)]
-        public virtual bool HasDropDownItems {
-            get {
+        public virtual bool HasDropDownItems
+        {
+            get
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs]:
```suggestion
            get => (dropDown != null) && dropDown.HasVisibleItems;
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs [hunk 21]
```diff
@@ -246,111 +287,131 @@ internal protected virtual Point DropDownLocation
...
-            get {
-                 // 
-                 if (dropDown != null) {
-                    if (DropDown.AutoClose || !IsInDesignMode || (IsInDesignMode && !IsOnDropDown)){
-                        return DropDown.OwnerItem == this && DropDown.Visible; 
+        public override bool Pressed
+        {
+            get
+            {
+                //
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs]: consider removing empty comment

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs [hunk 22]
```diff
@@ -246,111 +287,131 @@ internal protected virtual Point DropDownLocation
...
+            }
+
+            Point itemScreenLocation = this.TranslatePoint(Point.Empty, ToolStripPointType.ToolStripItemCoords, ToolStripPointType.ScreenCoords);
+            dropDownBounds.Location = new Point(itemScreenLocation.X + offset.X, itemScreenLocation.Y + offset.Y);
+            dropDownBounds = WindowsFormsUtils.ConstrainToScreenWorkingAreaBounds(dropDownBounds);
+            return dropDownBounds;
+        }
+
+
+        private void DropDown_Closed(object sender, ToolStripDropDownClosedEventArgs e)
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs]:
```suggestion
        private void DropDown_Closed(object sender, ToolStripDropDownClosedEventArgs e) =>  OnDropDownClosed(EventArgs.Empty);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs [hunk 23]
```diff
@@ -246,111 +287,131 @@ internal protected virtual Point DropDownLocation
...
+        {
             OnDropDownClosed(EventArgs.Empty);
         }
-   
-        private void DropDown_Opened(object sender, EventArgs e) {
+
+        private void DropDown_Opened(object sender, EventArgs e)
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs]:
```suggestion
        private void DropDown_Opened(object sender, EventArgs e) => OnDropDownOpened(EventArgs.Empty);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs [hunk 24]
```diff
@@ -246,111 +287,131 @@ internal protected virtual Point DropDownLocation
...
+        {
             OnDropDownOpened(EventArgs.Empty);
         }
 
-        private void DropDown_ItemClicked(object sender, ToolStripItemClickedEventArgs e) {
+        private void DropDown_ItemClicked(object sender, ToolStripItemClickedEventArgs e)
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs]:
```suggestion
        private void DropDown_ItemClicked(object sender, ToolStripItemClickedEventArgs e) => OnDropDownItemClicked(e);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs [hunk 25]
```diff
@@ -693,119 +814,202 @@ protected override void Dispose(bool disposing)
...
+                if (AccessibilityImprovements.Level1)
+                {
                     AccessibilityNotifyClients(AccessibleEvents.StateChange);
                     AccessibilityNotifyClients(AccessibleEvents.NameChange);
                 }
             }
         }
 
-        private bool ShouldSerializeDropDown() {
+        private bool ShouldSerializeDropDown()
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs]:
```suggestion
        private bool ShouldSerializeDropDown() => dropDown != null && !dropDown.IsAutoGenerated;
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs [hunk 26]
```diff
@@ -693,119 +814,202 @@ protected override void Dispose(bool disposing)
...
+        {
             return dropDown != null && !dropDown.IsAutoGenerated;
         }
 
-        private bool ShouldSerializeDropDownDirection() {
+        private bool ShouldSerializeDropDownDirection()
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs]:
```suggestion
        private bool ShouldSerializeDropDownDirection() => (toolStripDropDownDirection != ToolStripDropDownDirection.Default);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs [hunk 27]
```diff
@@ -693,119 +814,202 @@ protected override void Dispose(bool disposing)
...
+        {
             return (toolStripDropDownDirection != ToolStripDropDownDirection.Default);
         }
 
-        private bool ShouldSerializeDropDownItems() {
+        private bool ShouldSerializeDropDownItems()
```
Comment (JuditRose) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripDropDownItem.cs]:
```suggestion
        private bool ShouldSerializeDropDownItems() => (dropDown != null && dropDown.IsAutoGenerated);
```

**Comment by zsd4yr** [comment]
@KlausLoeffelmann is this still something we want to get in?

**Comment by zsd4yr** [comment]
@KlausLoeffelmann closing due to inactivity. Feel free to reopen when you have time 😄

**Comment by KlausLoeffelmann** [comment]
@merriemcgaw, @zsd4yr : I tested binaries again before merging, since merging was a bit complicated due to the many changes since I did the PR.

**Pull request merged**